java/ql/src/experimental/Security/CWE/CWE-078/CommandInjectionRuntimeExecTest.ql

```diff
@@ -0,0 +1,39 @@
+/**
```

Comment:
What's the reason to have these `testing`-labeled queries? If you want to add tests, you can do it under `java/ql/test/experimental`.

Reply:
The idea of those is to have a testing version of the query that is just a `problem` vs `path-problem` query, so that the testing is robust to minor changes in dataflow that don't change the outcome.

I've got tests, and wasn't sure where to put them. Thanks! I'll get them added there.

If those test versions live in the `test/experimental` directory, can they reference a test query that is local to them, and have that pick up the `.qll` used in the real queries, which are not in that testing directory?

Reply:
> The idea of those is to have a testing version of the query that is just a problem vs path-problem query, so that the testing is robust to minor changes in dataflow that don't change the outcome.

By changing the outcome, are you referring to alert wobble in Code Scanning? If that's the case, AFAIK alerts won't change as long as alert messages and primary locations (sinks in this case) are the same, even if the paths change slightly. I've never seen the need to have alternative `problem` versions of path queries, so happy to know more about your use case to see what can be done.

> If those test versions live in the test/experimental directory, can they reference a test query that is local to them, and have that pick up the .qll used in the real queries, which are not in that testing directory?

CodeQL files in the `test` directory can reference libraries under `lib`, but not queries under `src`. In any case, I think the main problem would be that queries in the `test/experimental` directory wouldn't be part of any CodeQL pack, so users wouldn't be able to use them in Code Scanning. If these are not really tests, but experimental user-oriented queries, then `src/experimental` is their place. But as I said above, we should really look into why these are necessary in the first place.

Reply:
We've seen a lot of cases where taint/dataflow paths change in other languages (especially Python) when the dataflow model changes, which makes tests fail frequently if they are path-sensitive, even without changes.

If that's not something that happens in Java then I can remove those and not worry about it!

I'll make that change and put the tests in.

Reply:
Ah, I see what you mean. That's with the old expectations test format, because edges and nodes vary. But if we use the inline expectations test format, that shouldn't be an issue anymore. I can help you with that if you're unfamiliar with it.

Reply:
See this for example: https://github.com/github/codeql/blob/b500bbba7660627bd31ec13fb70a13727e8b08b9/java/ql/test/query-tests/security/CWE-532/SensitiveLogInfo.ql

You just import `TaintFlowTest` parameterized with the configuration that you want to test (this requires the configuration to adhere to the new `DataFlow` module, see [the configuration for SQL injection](https://github.com/github/codeql/blob/b500bbba7660627bd31ec13fb70a13727e8b08b9/java/ql/lib/semmle/code/java/security/SqlInjectionQuery.qll#L39) as an example). Then, in the test code, you add a comment to indicate your expectation (either `// $ hasValueFlow` if it's a `DataFlow` configuration, or `// $ hasTaintFlow` if it's a `TaintTracking` configuration), and you can leave the `.expected` file mostly empty, with just the following contents:

```
failures
testFailures
```

That way the expectations are only in the source code, and as long as the sinks remain the same, tests won't break when paths change slightly.

If you don't want to refactor your dataflow configuration to the new format (although it only takes a few minutes), there's still a slightly more complex test `.ql` file that we can write to work with inline expectations. Let me know if you prefer this approach.

Reply:
Does that work with a query with more than one taint flow, or where we're doing something extra outside of the taint flow that's in the `select` clause of the query?

It looks like the test there is purely on the config, not on the entire query.

In the case of _this_ query it has two related sinks that are used in two _different_ taint flows. I'm not sure that fits into this `TaintFlowTest` model very well:

```codeql
from
  DataFlow::Node source, DataFlow::Node sink, ExecTaintConfiguration2 conf, MethodAccess call,
  int index, DataFlow::Node sourceCmd, DataFlow::Node sinkCmd, ExecTaintConfiguration confCmd
where
  call.getMethod() instanceof RuntimeExecMethod and
  // this is a command-accepting call to exec, e.g. exec("/bin/sh", ...)
  (
    confCmd.hasFlow(sourceCmd, sinkCmd) and
    sinkCmd.asExpr() = call.getArgument(0)
  ) and
  // it is tainted by untrusted user input
  (
    conf.hasFlow(source, sink) and
    sink.asExpr() = call.getArgument(index)
  )
select sink,
  "Call to dangerous java.lang.Runtime.exec() with command '$@' with arg from untrusted input '$@'",
  sourceCmd, sourceCmd.toString(), source, source.toString()
```

Reply:
I will still refactor the dataflow to use the new method, anyhow

Reply:
We'd need to add that logic to the test, which brings us to the complex variant (we use the raw `InlineExpectationsTest` library instead of `InlineFlowTest`). But it's still doable.

Reply:
Alternatively, you could add that logic to the definition of the sink of `ExecTaintConfiguration2`. Something like:

```ql
  override predicate isSink(DataFlow::Node sink) {
    exists(RuntimeExecMethod method, MethodAccess call |
      call.getMethod() = method and
      sink.asExpr() = call.getArgument(_) and // note that this includes the 0th argument as well, probably not what you want
      sink.asExpr().getType() instanceof Array and
      any(ExecTaintConfiguration c).hasFlowTo(DataFlow::exprNode(call.getArgument(0))
    )
  }
```

But you would lose the ability to reference the command in the alert message, unless you expose it somehow with another predicate.

Reply:
I'll get onto it on Thurs - thanks for your help so far with the better testing format!

Reply:
I updated to use the new dataflow with `ConfigSig`, and got a test working.

I also removed the test-specific versions of the queries in the `src` folder.

I stuck to the old test for now, since I wasn't sure how to write a test using `InlineExpectationsTest` yet.

Reply:
@atorralba - is this OK now?

I went with the old test type, since I wasn't sure how to use the inline version.

If you can walk me through doing the new way I'd be happy to make the change, or if you can accept the old test format that's fine by me too.

Reply:
Sorry for the delay. I added a few more comments.